### DISCUSSION TITLE
Date Picker

### DISCUSSION THREAD

**Discussion by claviska** [shoelace-style/webawesome#935] [Ideas & Suggestions]
See https://github.com/shoelace-style/shoelace/discussions/1653

- We can probably scope this to just a date picker. Time picker can come later and will be a separate component if we decide to pursue it.
- Date range functionality can come later too. It becomes significantly more complicated both in how form submission works and how accessibility is handled.

Prior art (web component-based):

- https://dia-date-picker.com/date-picker.html

I also created a calendar component prototype. We can take it or leave it, but the `calendar` branch has been updated. Preview here:

https://shoelace-git-calendar-font-awesome.vercel.app/components/calendar/

**Comment by gregjotau** [comment]
One important need we have is to have an input + datepicker. So input field which can expand into formatted date (shortcut based, custom), but to the right:

<img width="218" height="71" alt="image" src="https://github.com/user-attachments/assets/93cd6f3a-4f1d-4345-8781-77a76232f9bc" />

Is there any easy way to add a datepicker to the existing wa-input?

**Reply by RichLewis007** [reply]
@claviska Your calendar picker prototype looks great! What other calendar features are scoped? Is the accessibility aspect the most difficult in these components?

**Reply by claviska** [reply]
Accessibility is part of it, but also asking "what is a date picker?" Choosing a flight a month out is very different experience from scrolling through multiple decades to enter a birthdate, so I'd say the biggest challenge is accounting for the many use cases people have for inputing dates.

**Reply by RichLewis007** [reply]
Good points, thanks.

**Reply by dmrty** [reply]
@claviska I'll definitely NOT optimize the component for birthdate or any memorable dates. Gov.uk Design Team has a nice set of heavily-researched guidelines on that topic : 
https://design-system.service.gov.uk/patterns/dates/ (Read their GitHub for more)

**Comment by RichLewis007** [comment]
Hi,
 as a backer and advocate of Web Awesome and its idea of a quality framework‑agnostic library of open-source web components, I've been looking for ways to lend a hand. As a start, I've done a lot of research on data picker components, and have written up some guidance for the developers and community in selecting what capabilities it might best have.

First, as in all software engineering and especially in open source communities, let's be aware of what bike-shedding is, and consciously work to avoid it.

> Bikeshedding, also known as [Parkinson's Law of Triviality](https://en.wikipedia.org/wiki/Law_of_triviality), is the tendency for groups to focus disproportionately on minor, easily understood issues while neglecting more complex, important ones. This is commonly seen in open source forums and requirements definition processes because people naturally feel more comfortable talking about simple, familiar ideas rather than wrestling with more challenging, nuanced problems.

The following comments are a pragmatic, "don’t-overbuild it" blueprint we can begin with to help us converge on the right calendar/date-picker for Web Awesome.

**Comment by RichLewis007** [comment]
# 1) What to gather **before** writing requirements

As stakeholders of Web Awesome, let's work on answering these questions up front, as it prevents bikeshedding later. These help to frame requirements, and come together on scope, constraints and priorities so we don't overbuild the date picker (or spend too long bickering about features 😊).

Note: **APG** = W3C WAI-ARIA Authoring Practices Guide

* **Primary use cases (pick all that apply):**\
 ☐ **Single-date entry** - select one calendar day in a form (birthdate, due date, publish date).\
 ☐ **Date-range selection** - picking a start/end window (filters, validity periods, campaigns).\
 ☐ **Quick-pick ranges** - fast selection of common windows (Today, Yesterday, Last 7/30, This/Last Month/Quarter).\
 ☐ **Inline calendar browsing** - always-visible calendar (dashboards, side panels).\
 ☐ **Multiple discrete dates** (optional) - select several non-contiguous days (exceptions, blackout dates).\
 ☐ **Mobile/touch-first mode** - full-screen or bottom-sheet presentation optimized for touch.\
 ☐ **Accessibility-first workflows** - keyboard-only and screen-reader for high-compliance contexts.

* **Interaction model(s) needed**:\
  ☐ text input with popup\
  ☐ inline calendar\
  ☐ modal dialog\
  ☐ mobile full-screen\
  ☐ all of the above

* **Granularity**:\
  ☐ date only\
  ☐ date-range
  
* **Internationalization**:\
  ☐ locales\
  ☐ first-day-of-week\
  ☐ week numbers\
  ☐ 12/24h preferences\
  ☐ RTL\
  ☐ which calendars (Gregorian only for v1?)

* **Validation**:\
  ☐ min/max\
  ☐ disabled rules (weekends/holidays/functions)\
  ☐ required\
  ☐ parsing/formatting rules\
  ☐ free-typing policy

* **Performance constraints**:\
  ☐ must work smoothly in forms with 100+ instances\
  ☐ SSR (Server Side Rendering) (can't get these bullets to format properly..)\
  -☐ make it SSR-safe and progressively enhanced\
  -☐ don’t break when users do SSR/SSG; degrade gracefully before JS loads\
  -☐ this could be in scope even though we are framework-neutral\
  ☐ shadow DOM constraints?


* **Design tokens & theming**:\
  ☐ which tokens (radius, spacing, color roles, focus rings) must be consumable\
  ☐ dark mode\
  ☐ high-contrast mode

* **Accessibility bar (W3C)**:\
  ☐ target WCAG 2.1 AA (accessability) https://www.levelaccess.com/resources/must-have-wcag-2-1-checklist/\
  ☐ follow APG patterns (dialog + grid or combobox pattern)

* **API shape**:\
  ☐ properties\
  ☐ events\
  ☐ slots/parts\
  ☐ controlled vs uncontrolled value\
  ☐ formatting hooks\
  ☐ keyboard map

* **Ecosystem fit**:\
  ☐ compose with Web Awesome inputs\
  ☐ popovers\
  ☐ buttons\
  ☐ icons\
  ☐ form field\
  ☐ localization service

* **Testing matrix**:\
  ☐ keyboard/screen reader (NVDA, JAWS, VoiceOver)\
  ☐ RTL\
  ☐ mobile (iOS/Android)\
  ☐ high-contrast

**Out-of-Scope for this Component:**
  ☐ **date-time composition** (optional) - compose with a time picker; unified commit semantics.\
  ☐ date-time (and whether time belongs in a separate component that can compose with the calendar)
"Time picker can come later and will be a separate component if we decided to pursue it."

**Comment by RichLewis007** [comment]
# 2) Date Picker Landscape scan - how others do it (highlights)

  - WAI-ARIA Authoring Practices document two canonical approaches: a date-picker dialog with grid navigation and a combobox that opens a dialog. Both define concrete keyboard behavior (arrows, PgUp/PgDn to change months/years, Esc to close, etc.). This is the north star for a11y. https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/datepicker-dialog/

  - Open UI’s Date Picker research aggregates behaviors across major systems; useful as a neutral spec reference when bikeshedding. https://open-ui.org/components/datepicker.research/

  - Material UI (MUI X): Desktop vs Mobile variants, strong a11y and keyboard coverage, slot-based subcomponents for extensibility, validation API, and timezone support. Good model for a "composable but opinionated" design. https://mui.com/x/react-date-pickers/date-picker/

  - Ant Design: rich picker modes (date/month/quarter/year), confirm-button option (needConfirm) that alters commit semantics. https://ant.design/components/date-picker

  - Flatpickr (vanilla JS): small, fast, feature-dense: range mode, multiple months, enable/disable via functions, custom formatter/hooks. Useful benchmark for options density. https://flatpickr.js.org/options/

  - PrimeReact: multiple selection modes (single/multiple/range) and explicit keyboard maps in docs; a good public reference for keyboard expectations. https://www.primefaces.org/primereact-v8/calendar/ and https://v9.primereact.org/calendar/

  - Kendo/Syncfusion (paid): deep enterprise features (decade navigation depth, templates, date-time and range variants, theming bundles). They set the "power-user" bar. https://www.telerik.com/kendo-jquery-ui/documentation/controls/datepicker/overview and https://demos.telerik.com/kendo-ui/datepicker/index

  - FullCalendar (scheduler, not a field widget): shows advanced selection behaviors and API semantics for ranges — relevant if WA ever adds a "big calendar". https://fullcalendar.io/docs

  - Carbon (IBM) provides a11y guidance you can mirror for docs. https://carbondesignsystem.com/components/date-picker/accessibility/

**Comment by RichLewis007** [comment]
# 3) Master list of possible features (pick what matters)

Note: APG = W3C WAI-ARIA Authoring Practices Guide

## Core UX & Interaction

  - Text input + popup calendar. Inline mode. Modal/dialog mode.
  - Keyboard support (full APG grid + dialog): Arrows to move day. PgUp/PgDn month. Shift+PgUp/PgDn year. Home/End row. Esc close. Enter select.
  - Mouse & touch friendly targets. Focus trapping in dialog. Typeahead for month/year.
  - Controlled/uncontrolled value. Open/close programmatically.

## Selection modes

  - Single date.
  - **Range** (with hover-preview between start & end)
  - Multiple dates (optional. less common)
  - **Month**, **Year**, **Decade** pickers (jump navigation)

## Validation & rules

  - Min/max, disabled days, disable functions (e.g., "no weekends/holidays")
  - Block out arbitrary sets ('enable'/'disable' by list, ranges, function)
  - Commit semantics: immediate vs "confirm" button.
  - Free typing with strict parsing + error states (and clear button)

## Calendaring & display

  - Multi-month view (1-n columns), month dropdown, year dropdown.
  - First day of week; week numbers toggle; today button; clear button. (Week numbers are explicitly referenced in several systems)
  - Show holidays/markers via custom render/slots.

## Formatting & i18n

  - Locale, numbering system, calendar system (Start with **Gregorian**. Consider hooks for others later)
  - Custom input/output formats. Alt display format vs machine format.
  - RTL support. Alternate day header formats (short/narrow/wide)

## Theming & customization

  - Design tokens: radius, gap, spacing, stroke/focus, colors (bg, text, hover, selected, range-preview, disabled)
  - CSS Parts/slots for header, nav buttons, month grid, day cells, range start/end, outside-month days.
  - Icon slots (prev/next/today/clear). Action bar slot.

## API & events

  - API: 'value', 'defaultValue', 'open', 'min', 'max', 'disabledDates', 'firstDayOfWeek', 'showWeekNumbers', 'numberOfMonths', 'selectionMode', 'locale', 'dir', 'confirm'/'commitOnClose', 'presets=[{label, range|date}]', 'renderDay' (templating), 'format', 'parse'.
  - Events: 'input', 'change', 'open', 'close', 'month-change', 'year-change', 'invalid', 'confirm', 'range-hover'.
  - Flatpickr’s event richness is a good reference. https://github.com/flatpickr/flatpickr

## Accessibility

  - Role/ARIA alignment per APG. Labelled fields and instructions. Live region for month/year changes. High contrast and focus rings. Screen reader announcements for range start/end.
  - Combobox pattern parity for input variant.

## Nice quality-of-life extras

  - Preset ranges (Today, Yesterday, Last 7/30, This/Last month/quarter)
  - "Keep open after selection" and "close on range selection" toggles.
  - Auto-apply on blur vs require explicit confirm (configurable)
  - Multiple calendars (Buddhist, ISO week date, etc.) **advanced**.
  - Virtualized year/month lists for perf on low-power devices.
  - Holiday providers hook.

**Out-of-scope for this Component**:
  - Timezone awareness (if composing with a time picker / date-time)

**Comment by RichLewis007** [comment]
# 4) Prioritize: Essential vs "most wanted" vs "nice-to-have"

**Suggested Essential (v1 "must ship")**

  - APG-compliant (W3C WAI-ARIA Authoring Practices Guide) keyboard & screen-reader behavior (dialog and/or combobox patterns).
  - Single date **and** range selection.
  - Min/max & disable rules (function-based)
  - Multi-month (2) option, Month/Year fast nav.
  - Locale + first-day-of-week. Formatting/parsing. RTL (Right-to-Left locales)
  - Theming via tokens + CSS parts. Dark mode. High-contrast.
  - Events for 'change/open/close/month-change/year-change'.
  - Form integration: validation states, clear button, today shortcut.

**Suggested Most used / high impact (v1.1-v2)**

  - Preset ranges. Hover range preview. Week numbers toggle.
  - Commit semantics (confirm bar) for Enterprise forms.
  - Inline mode + mobile-optimized modal variant.
